bazel/external/lightstep.BUILD

```diff
@@ -0,0 +1,73 @@
+load("@//bazel/external:helpers.bzl", "cc_genrule_deps")
+
+cc_library(
+    name = "lightstep",
+    srcs = [":_prefix/lib/liblightstep_core_cxx11.a"],
+    hdrs = glob([
+        "src/c++11/lightstep/**/*.h",
+        "src/c++11/mapbox_variant/**/*.hpp",
+    ]) + [
+        ":_prefix/include/collector.pb.h",
+        ":_prefix/include/lightstep_carrier.pb.h",
+    ],
+    includes = ["src/c++11", "_prefix/include"],
+    deps = ["@protobuf_bzl//:protobuf"],
+    visibility = ["//visibility:public"],
+)
+
+cc_genrule_deps(
+    name = "protobuf_deps",
+    deps = ["@protobuf_bzl//:protobuf"],
+)
+
+genrule(
+    name = "install",
+    srcs = glob(["**"]) + [
+        ":protobuf_deps",
+        "@protobuf_bzl//:well_known_protos",
+    ],
+    cmd = """
+    configure="$${PWD}/$(location :configure)"
+    protoc="$${PWD}/$(location @protobuf_bzl//:protoc)"
+    protobuf_src="$${PWD}/external/protobuf_bzl/src/"
+    protobuf_libdir="$${PWD}/$(BINDIR)/external/protobuf_bzl/"
+    mkdir "$(@D)/build"
+
+    # see https://github.com/envoyproxy/envoy/issues/1387 for progress
+    cp -r "$$(dirname $${configure})" "$(@D)/src"
+    cd "$(@D)/src"
+    cat > ../lightstep-missing-header.diff << EOF
+--- ./src/c++11/lightstep/options.h.bak        2017-08-04 09:30:19.527076744 -0400
++++ ./src/c++11/lightstep/options.h    2017-08-04 09:30:33.742106924 -0400
+@@ -5,6 +5,7 @@
+ // Options for Tracer implementations, starting Spans, and finishing
+ // Spans.
+ 
++#include <functional>
+ #include <chrono>
+ #include <memory>
+ 
+EOF
+    (cd src/c++11/lightstep/ && cat options.h > options.h.new && mv options.h.new options.h)
+    patch -p0 < ../lightstep-missing-header.diff
+
+    cd "../build"
+    ../src/configure \
+        --prefix="$${PWD}/../_prefix" \
+        --disable-grpc \
+        --enable-shared=no \
+        protobuf_CFLAGS="-I$${protobuf_src}" \
+        protobuf_LIBS="-L$${protobuf_libdir} -lprotobuf -lprotobuf_lite" \
+        PKG_CONFIG="true" \
+        PROTOC="$${protoc} -I$${protobuf_src}"
+    make V=1 install
+    """,
+    outs = [
+        "_prefix/lib/liblightstep_core_cxx11.a",
+        "_prefix/include/collector.pb.h",
+        "_prefix/include/lightstep_carrier.pb.h",
```

Comment:
This is one of the reasons we opted for using a `repository_rule` instead of `genrule`. You have to manually enumerate and maintain the complete list of outputs in `genrules`. For simple cases it's OK, but when you have a large set of headers, which is the case in some libs, it becomes really big.

Reply:
ACK. I _think_ it should be possible to avoid huge manual enumerations by using `glob()`, since it's uncommon for large numbers of output files to use arbitrary names. For example, a library that installed ~dozens of proto headers could use `glob("**/*.proto")` to derive the output names.

Reply:
I don't think genrules support globs on outputs FYI, at least last time I checked. Maybe this has changed?

Reply:
See https://github.com/jmillikin/bazel-packages/blob/master/libevent/libevent.bzl for an example. The glob is on input files, and used to calculate the output file names using normal Python expressions.